primary/src/grpc_server/validator.rs

```diff
@@ -2,40 +2,89 @@
 // SPDX-License-Identifier: Apache-2.0
 use std::time::Duration;
 
-use crate::{block_waiter::GetBlockResponse, BlockCommand};
+use crate::{block_waiter::GetBlockResponse, BlockCommand, BlockRemoverCommand};
 use tokio::{
-    sync::{mpsc::Sender, oneshot},
+    sync::{mpsc::channel, mpsc::Sender, oneshot},
     time::timeout,
 };
 use tonic::{Request, Response, Status};
 use types::{
-    BatchMessageProto, BlockError, CollectionRetrievalResult, GetCollectionsRequest,
+    BatchMessageProto, BlockError, CollectionRetrievalResult, Collections, Empty,
     GetCollectionsResponse, Validator,
 };
 
 #[derive(Debug)]
 pub struct NarwhalValidator {
     tx_get_block_commands: Sender<BlockCommand>,
     get_collections_timeout: Duration,
+    tx_block_removal_commands: Sender<BlockRemoverCommand>,
+    remove_collections_timeout: Duration,
 }
 
 impl NarwhalValidator {
     pub fn new(
         tx_get_block_commands: Sender<BlockCommand>,
         get_collections_timeout: Duration,
+        tx_block_removal_commands: Sender<BlockRemoverCommand>,
+        remove_collections_timeout: Duration,
     ) -> Self {
         Self {
             tx_get_block_commands,
             get_collections_timeout,
+            tx_block_removal_commands,
+            remove_collections_timeout,
         }
     }
 }
 
 #[tonic::async_trait]
 impl Validator for NarwhalValidator {
+    async fn remove_collections(
+        &self,
+        request: Request<Collections>,
+    ) -> Result<Response<Empty>, Status> {
+        let collection_ids = request.into_inner().collection_ids;
+        let remove_collections_response = if !collection_ids.is_empty() {
+            let (tx_remove_block, mut rx_remove_block) = channel(1);
+            let mut ids = vec![];
+            for collection_id in collection_ids {
+                ids.push(collection_id.try_into().map_err(|err| {
+                    Status::invalid_argument(format!("Could not serialize: {:?}", err))
+                })?);
+            }
+            self.tx_block_removal_commands
+                .send(BlockRemoverCommand::RemoveBlocks {
+                    ids,
+                    sender: tx_remove_block,
+                })
+                .await
+                .map_err(|err| Status::internal(format!("Send Error: {:?}", err)))?;
```

Comment:
If I understand correctly this message will be visible to the consumer of the API. Maybe it won't make much sense to expose the "internal" errors (e.x couldn't send a message to a tokio channel) to the client? We could come up with something generic. Not an issue for now, but just highlighting.

Reply:
Hmmm... yeah this might require some more thought. But I will leave the error for now as we are building and maybe go back and make the errors more generic as to not expose too much to the customer.

Reply:
If you could open an issue for this it would be great so we make sure we track it

Reply:
Opened [Issue#234](https://github.com/MystenLabs/narwhal/issues/234)